src/classes/CCON_PreventUpdate_TDTM.cls

```diff
@@ -0,0 +1,72 @@
+/*
+    Copyright (c) 2019, Salesforce.org
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Course Enrollments
+* @group-content ../../ApexDocContent/CourseEnrollments.htm
+* @description When a Term Grade or an Attendance Event is present on a Course Connection, 
+* this class prevents changes to a Contact or Course Offering on that Course Connection.
+*/
+public with sharing class CCON_PreventUpdate_TDTM extends TDTM_Runnable {
+   
+    /*******************************************************************************************************
+    * @description When a Term Grade or an Attendace Event is present on a Course Connection,
+    * this class prevents changes to a Contact or Course Offering fields on the record.
+    * @param listNew the list of Course Connections from trigger new.
+    * @param listOld the list of Course Connections from trigger old.
+    * @param triggerAction which trigger event (BeforeUpdate).
+    * @param objResult the describe for Course Connections.
+    * @return dmlWrapper
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist,
+    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+    
+        Map<ID, sObject> newMap = new Map<ID, sObject>(newList);
```

Comment:
This whole logic looks fantastic. The only thing I might recommend is that I think this could be `Map<Id, Course_Enrollment__c> newMap = new Map<Id, Course_Enrollment__c>(newList);` (you might need to do some casting here, I'm not sure)

Reply:
At the very least, it would allow you to clean up the code on lines 57 & 58 - you won't need to do the (Course_Enrollment__c) casting there.

Reply:
The casting might require another for loop so I think this is ok, lmk.

Reply:
See updated comment above - I think this can be done with some casting.